### DISCUSSION TITLE
How to get project endpoint addresses to configure other projects

### DISCUSSION THREAD

**Discussion by paulomorgado** [microsoft/aspire#1560] [Q&A]
In my Aspire playground I have a very simple identity provider that is used by the other projects.

ASP.NET Core 8.0 JWT bearer infrastructure does not use service discovery and I don't want to force it into the projects, so I'm setting it in Aspire:

```csharp
var idp = builder.AddProject<Projects.Identitty>("Identity")
    .WithLaunchProfile("aspire")
    ;

var webService = builder.AddProject<Projects.WebService>("WebService")
    .WithLaunchProfile("aspire")
    .WithEnvironment("Security__Authentication__JwtBearerOptions__Authority", $"http://localhost:5501")
    .WithEnvironment("Security__Authentication__JwtBearerOptions__MetadataAddress", $"http://localhost:5501/.well-known/openid-configuration")
    ;
```

How can I get the endpoint for `idp` to set in `webService`?

**Comment by paulomorgado** [comment]
@davidfowl [answer on another question](https://github.com/dotnet/aspire/discussions/1142#discussioncomment-7720575) hints on a solution for the simple case of `Security__Authentication__JwtBearerOptions__Authority`, but doesn't solve for `Security__Authentication__JwtBearerOptions__MetadataAddress`-

**Reply by davidfowl** [reply]
idp.GetEndpoint("http") ?

https://github.com/dotnet/eShop/blob/63e82c9e75b1a50cd28499240fcbc2942bdc6939/src/eShop.AppHost/Program.cs#L62

**Reply by paulomorgado** [reply]
It doesn't work if I have to compose it, like in the example above.

**Reply by paulomorgado** [reply]
How can I use service discovery on a project where I know I have components that are not yet ready for that?

**Reply by DamianEdwards** [reply]
Service Discovery should work for any component where the `HttpClient` is exposed. If the `HttpClient` is internal to the component, then you'd need to manually resolve the address using the Service Discovery APIs at application startup and then configure the component to use the resolved address, but of course then you won't get any on-demand features of Service Discovery (e.g. client-side load balancing, etc.).

**Reply by paulomorgado** [reply]
That's the case of authentication and authorization.

For now, I would be happy with just getting the first address from service discovery to override the values in configuration.

Are there any docs or samples on how to do that?

**Comment by paulomorgado** [comment]
For `JwtBearer` I'm doing this:

```csharp
services
    .AddAuthentication()
    .AddJwtBearer();

services.ConfigureOptions<JwtBearerPostConfigureOptions>();

// ...

private sealed class JwtBearerPostConfigureOptions : IPostConfigureOptions<JwtBearerOptions>
{
    private readonly IHttpClientFactory _httpClientFactory;

    public JwtBearerPostConfigureOptions(IHttpClientFactory httpClientFactory)
    {
        _httpClientFactory = httpClientFactory;
    }
    public void PostConfigure(string? name, JwtBearerOptions options)
    {
        options.Backchannel = string.IsNullOrEmpty(name) ? _httpClientFactory.CreateClient() : _httpClientFactory.CreateClient(name);
    }
}
```

This way, `JwtBearer` not only can use service discovery, but can also use resilience.
